distributed/tests/test_resources.py

```diff
@@ -261,3 +261,20 @@ def test_dont_optimize_out(c, s, a, b):
 
     for key in map(tokey, y.__dask_keys__()):
         assert 'executing' in str(a.story(key))
+
+
+@gen_cluster(client=True, ncores=[('127.0.0.1', 1, {'resources': {'A': 1}}),
+                                  ('127.0.0.1', 1, {'resources': {'B': 1}})])
+def test_compute_multidim(c, s, a, b):
+    da = pytest.importorskip('dask.array')
+    np = pytest.importorskip('numpy')
+    x = delayed(np.random.randint)(0, 1, (5,5))
+    y = da.from_delayed(x, (5,5), int)
+
+    xx = c.compute(x, resources={x: {'A': 1}},)
+    yy = c.compute(y, resources={y: {'A': 1}},)
```

Comment:
Looking at the test failures in travis-ci it seems like this is illegal because `y` is not hashable?

```python
    @gen_cluster(client=True, ncores=[('127.0.0.1', 1, {'resources': {'A': 1}}),
                                      ('127.0.0.1', 1, {'resources': {'B': 1}})])
    def test_compute_multidim(c, s, a, b):
        da = pytest.importorskip('dask.array')
        np = pytest.importorskip('numpy')
        x = delayed(np.random.randint)(0, 1, (5,5))
        y = da.from_delayed(x, (5,5), int)
    
        xx = c.compute(x, resources={x: {'A': 1}},)
>       yy = c.compute(y, resources={y: {'A': 1}},)
E       TypeError: unhashable type: 'Array'
```

Reply:
That is not surprising, but the weird thing is that the travis-ci test fails in a different way. The travis-ci test fails in a way that I can reproduce locally, but I don't understand `data` property of the worker.
I've removed the assertion related to dask.array. It passes locally now.

Reply:
That assertion was the main reason for this PR.  We genuinely can't pass `resources={x: ...}` where x is a dask array because dask arrays are not hashable.  The flatten logic you've added in this PR probably needs to be called on the user side.  Alternatively, we could introduce some sort of list of key-value-pairs alternative input to `resources=` to handle non-hashable keys.